internal/core/ebook.go

```diff
@@ -0,0 +1,238 @@
+package core
+
+import (
+	"archive/zip"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"os"
+	fp "path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/go-shiori/shiori/internal/model"
+)
+
+func EbookGenerate(req ProcessRequest) (book model.Bookmark, isFatalErr bool, err error) {
+	// variable for store generated html code
+	var html string
+
+	book = req.Bookmark
+
+	// Make sure bookmark ID is defined
+	if book.ID == 0 {
+		return book, true, fmt.Errorf("bookmark ID is not valid")
+	}
+
+	// cheak archive and thumb
+	strID := strconv.Itoa(book.ID)
+
+	imagePath := fp.Join(req.DataDir, "thumb", fmt.Sprintf("%d", book.ID))
+	archivePath := fp.Join(req.DataDir, "archive", fmt.Sprintf("%d", book.ID))
+
+	if _, err := os.Stat(imagePath); err == nil {
+		book.ImageURL = fp.Join("/", "bookmark", strID, "thumb")
+	}
+
+	if _, err := os.Stat(archivePath); err == nil {
+		book.HasArchive = true
+	}
+
+	if _, err := os.Stat(archivePath); err == nil {
+		book.HasArchive = true
+	}
+	ebookPath := fp.Join(req.DataDir, "ebook", fmt.Sprintf("%d.epub", book.ID))
+	// if epub exist finish prosess else continue
+	if _, err := os.Stat(ebookPath); err == nil {
+		return book, false, nil
+	}
+
+	ebookDir := fp.Join(req.DataDir, "ebook")
+	// check if directory not exsist create that
+	if _, err := os.Stat(ebookDir); os.IsNotExist(err) {
+		os.MkdirAll(ebookDir, model.DataDirPerm)
+	}
+	// create epub file
+	epubFile, err := os.Create(ebookPath)
+	if err != nil {
+		return book, true, fmt.Errorf("can't create ebook")
+	}
+	defer epubFile.Close()
+
+	// Create zip archive
+	epubWriter := zip.NewWriter(epubFile)
+	defer epubWriter.Close()
+
+	// Create the mimetype file
+	mimetypeWriter, err := epubWriter.Create("mimetype")
+	if err != nil {
+		return book, true, fmt.Errorf("can't create mimetype")
+	}
+	mimetypeWriter.Write([]byte("application/epub+zip"))
+
+	// Create the container.xml file
+	containerWriter, err := epubWriter.Create("META-INF/container.xml")
+	if err != nil {
+		return book, true, fmt.Errorf("can't create container.xml")
+	}
+
+	containerWriter.Write([]byte(`<?xml version="1.0" encoding="UTF-8"?>
+<container version="1.0" xmlns="urn:oasis:names:tc:opendocument:xmlns:container">
+  <rootfiles>
+	<rootfile full-path="OEBPS/content.opf" media-type="application/oebps-package+xml"/>
+  </rootfiles>
+</container>`))
+
+	contentOpfWriter, err := epubWriter.Create("OEBPS/content.opf")
+	if err != nil {
+		return book, true, fmt.Errorf("can't create content.opf")
+	}
+	contentOpfWriter.Write([]byte(`<?xml version="1.0" encoding="UTF-8"?>
+<package xmlns="http://www.idpf.org/2007/opf" version="2.0" unique-identifier="BookId">
+  <metadata>
+    <dc:title>` + book.Title + `</dc:title>
+    <dc:language>en</dc:language>
+    <dc:identifier id="BookId">urn:uuid:12345678-1234-5678-1234-567812345678</dc:identifier>
```

Comment:
Should we create an identifier for each bookmark? What does this imply?

Reply:
If the field is not required or we don't know the usage, we can avoid it for now. Keep me posted 👍